HTMLSerializer.js

```diff
@@ -0,0 +1,326 @@
+   *     attribute.
+   * @private
+   */
+  processSrcHole(element) {
+    var win = element.ownerDocument.defaultView;
+    var src = element.attributes.src;
+    var quote = this.escapedQuote(this.windowDepth(win));
+    this.html.push(`${src.name}=${quote}`);
+    this.srcHoles[this.html.length] = this.fullyQualifiedURL(element).href;
+    this.html.push(''); // Entry where data url will go.
+    this.html.push(quote + ' '); // Add a space before the next attribute.
+  }
+
+  /**
+   * Add a name and value pair to the list of attributes in |this.html|.
+   *
+   * @param {Window} win The window of the Element that is being processed.
+   * @param {string} name The name of the attribute.
+   * @param {string} value The value of the attribute.
+   */
+  processSimpleAttribute(win, name, value) {
+    var quote = this.escapedQuote(this.windowDepth(win));
+    this.html.push(`${name}=${quote}${value}${quote} `);
+  }
+
+  /**
+   * Computes the index of the window in its parent's array of frames.
+   *
+   * @param {Window} childWindow The window to use in the calculation.
+   * @return {number} the frames index.
+   */
+  iframeIndex(childWindow) {
+    if (childWindow.parent != childWindow) {
+      for (var i = 0; i < childWindow.parent.frames.length; i++) {
+        if (childWindow.parent.frames[i] == childWindow) {
+          return i;
+        }
+      }
+    } else {
+      return -1;
+    }
+  }
+
+  /**
+   * Computes the full path of the frame in the root document. Nested layers
+   * are seperated by '.'.
+   *
+   * @param {Window} win The window to use in the calculation.
+   * @return {string} The full path.
+   */
+  iframeFullyQualifiedName(win) {
+    if (this.iframeIndex(win) < 0) {
+      return '0';
+    } else {
+      var fullyQualifiedName = this.iframeFullyQualifiedName(win.parent);
+      var index = this.iframeIndex(win);
+      return fullyQualifiedName + '.' + index; 
+    }
+  }
+
+  /**
+   * Calculate the correct quotes that should be used given the nesting depth of
+   * the window in the frame tree.
+   *
+   * @param {number} depth The nesting depth of this window in the frame tree.
+   * @return {string} The correctly escaped quotation marks.
+   */
+  escapedQuote(depth) {
+    if (depth == 0) {
+      return '"';
+    } else {
+      return '&' + new Array(depth).join('amp;') + 'quot;';
+    }
+  }
+
+  /**
+   * Calculate the nesting depth of a window in the frame tree.
+   *
+   * @param {Window} win The window to use in the calculation.
+   * @return {number} The nesting depth of the window in the frame trees.
+   */
+  windowDepth(win) {
+    return this.iframeFullyQualifiedName(win).split('.').length - 1;
+  }
+
+  /**
+   * Take all of the srcHoles and create data urls for the resources, placing
+   * them in |this.html|. Calls the callback when complete.
+   *
+   * @param {Function} callback The callback function.
+   */
+  fillSrcHoles(callback) {
```

Comment:
I had initially, kept fillSrcHoles out of HTMLSerializer because it's behavior is asynchronous, and everything else in HTMLSerializer is synchronous, but I supposed that's no reason to keep it out.  But I think this results in the comment at the top of HTMLSerializer being slightly incorrect.